pkg/scheduler/framework/plugins/podtopologyspread/scoring.go

```diff
@@ -0,0 +1,293 @@
+				cnt = *s.TopologyPairToPodCounts[pair]
+			}
+			score += scoreForCount(cnt, c.MaxSkew, s.TopologyNormalizingWeight[i])
+		}
+	}
+	return int64(score), nil
+}
+
+// NormalizeScore invoked after scoring all nodes.
+func (pl *PodTopologySpread) NormalizeScore(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, scores framework.NodeScoreList) *framework.Status {
+	s, err := getPreScoreState(cycleState)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	if s == nil {
+		return nil
+	}
+
+	// Calculate <minScore> and <maxScore>
+	var minScore int64 = math.MaxInt64
+	var maxScore int64
+	for _, score := range scores {
+		// it's mandatory to check if <score.Name> is present in m.IgnoredNodes
+		if s.IgnoredNodes.Has(score.Name) {
+			continue
+		}
+		if score.Score < minScore {
+			minScore = score.Score
+		}
+		if score.Score > maxScore {
+			maxScore = score.Score
+		}
+	}
+
+	for i := range scores {
+		nodeInfo, err := pl.sharedLister.NodeInfos().Get(scores[i].Name)
+		if err != nil {
+			return framework.NewStatus(framework.Error, err.Error())
+		}
+		node := nodeInfo.GetNode()
+
+		if s.IgnoredNodes.Has(node.Name) {
+			scores[i].Score = 0
+			continue
+		}
+
+		if maxScore == 0 {
+			scores[i].Score = framework.MaxNodeScore
+			continue
+		}
+
+		s := scores[i].Score
+		scores[i].Score = framework.MaxNodeScore * (maxScore + minScore - s) / maxScore
+	}
+	return nil
+}
+
+// ScoreExtensions of the Score plugin.
+func (pl *PodTopologySpread) ScoreExtensions() framework.ScoreExtensions {
+	return pl
+}
+
+func getPreScoreState(cycleState *framework.CycleState) (*preScoreState, error) {
+	c, err := cycleState.Read(preScoreStateKey)
+	if err != nil {
+		return nil, fmt.Errorf("error reading %q from cycleState: %v", preScoreStateKey, err)
+	}
+
+	s, ok := c.(*preScoreState)
+	if !ok {
+		return nil, fmt.Errorf("%+v  convert to podtopologyspread.preScoreState error", c)
+	}
+	return s, nil
+}
+
+// topologyNormalizingWeight calculates the weight for the topology, based on
+// the number of values that exist for a topology.
+// Since <size> is at least 1 (all nodes that passed the Filters are in the
+// same topology), and k8s supports 5k nodes, the result is in the interval
+// <1.09, 8.52>.
+//
+// Note: <size> could also be zero when no nodes have the required topologies,
+// however we don't care about topology weight in this case as we return a 0
+// score for all nodes.
+func topologyNormalizingWeight(size int) float64 {
+	return math.Log(float64(size + 2))
+}
+
+// scoreForCount calculates the score based on number of matching pods in a
+// topology domain, the constraint's maxSkew and the topology weight.
+// `maxSkew-1` is added to the score so that differences between topology
+// domains get watered down, controlling the tolerance of the score to skews.
+func scoreForCount(cnt int64, maxSkew int32, tpWeight float64) float64 {
+	return float64(cnt)*tpWeight + float64(maxSkew-1)
```

Comment:
What are the parameters `maxSkew` and `tpWeight` for? What will be the difference if these two parameters are removed (i.e. score = cnt)?

Reply:
This is the original calculation method of [k8s](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/podtopologyspread/scoring.go#L303). I am not very clear about why it is calculated in this way. I can try to explain it as follows:
* `maxSkew`: represents the maximum imbalance value allowed by this constraint, `maxSkew = the maximum number of pods in a topology domain - the minimum number of pods in a topology domain`
* `tpWeight`: represents the number of filtered nodes `tpWeight = (log(len(filterNodes+2)))`
* `cnt`: the number of matched pods

In this calculation formula, `float64(cnt)*tpWeight` shows that when the number of pods is the same, it is more inclined to schedule to the topology domain with **more nodes**, and `+float64(maxSkew-1)` in the formula is a small factor, which only represents the difference between different constraints to a certain extent, it also can make the scheduling **more **stable** when the pods scheduled at the beginning are very small